### ISSUE TITLE
Eval bug: vulkan flash attention lets stale K/V in freed cells influence the output

### ISSUE THREAD

**Issue by natemacfadden** [ggml-org/llama.cpp#26744]
### Name and Version

version: 10319 (4cb22cd53)
built with GNU 15.2.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

Vulkan

### Hardware

AMD RYZEN AI MAX+ 395 w/ Radeon 8060S

122 GB

Available devices:
  Vulkan0: Radeon 8060S Graphics (RADV STRIX_HALO) (127488 MiB, 127196 MiB free)

### Models

Should be reproducible on any model I think (be careful with models that have a recurrent state... see my other issue https://github.com/ggml-org/llama.cpp/issues/26695). I used
[bartowski/Qwen2.5-0.5B-Instruct-GGUF](https://huggingface.co/bartowski/Qwen2.5-0.5B-Instruct-GGUF)

### Problem description & steps to reproduce

(This was found with assistance by Claude Opus 5; I reviewed and validated the bug)

When you rollback >1 tokens, the KV cache slots for tokens are not cleared (just masked away with -Inf before softmax). This *should* have no effect because exp(-inf + x) = +0 for any finite x, the stale slot's attention score. So, as long as V is finite (otherwise `0*Inf = NaN`), IEEE 754-2019 §7.2, §6.3 suggests the attention sum would be unchanged (`+0*V = +/- 0`; no effect if the computations are IEEE conformant). As the code shows below, however, the masked-out slots do have a (very small; unless V is ever Inf in which case it is a large effect... corrupting a sum) effect.

This bug does NOT occur if either `GGML_VK_DISABLE_COOPMAT=1` or the computation is done on the CPU. This suggests a shader issue. I believe, specifically, it is [this shader](https://github.com/ggml-org/llama.cpp/blob/4cb22cd537a9b12b717bd725b3fecb83c94894eb/ggml/src/ggml-vulkan/vulkan-shaders/flash_attn_cm1.comp#L479-L496) lines 479-496 specifically `PVMat = coopMatMulAdd(KMat, QMat, PVMat);`. This can be fixed by zeroing fully-masked K/V columns while staging the tile.

This looks like one of the mechanisms behind [#25618](https://github.com/ggml-org/llama.cpp/issues/25618). It is independent of the GQA-packing problem being fixed in [#26358](https://github.com/ggml-org/llama.cpp/pull/26358).

(The following test script was originally written under my direction by Claude Opus 5, but I read/understood each line and modified it for brevity and clarity)

```
// build from the llama.cpp root, after cmake --build build:
// g++ -O2 -std=c++17 fa_stale_kv_leak.cpp -o build/bin/fa_stale_kv_leak -Iinclude -Iggml/include \
//     -Lbuild/bin -lllama -lggml -lggml-base -Wl,-rpath,'$ORIGIN'
// the binary must land in build/bin: backends load from the executable's own directory

#include "llama.h"
#include "ggml-backend.h"
#include <cstdio>
#include <vector>

static const int WARM = 12, N = 2;   // history; block to decode-then-remove (1 freed cell keeps its K/V)
static llama_model * model;
static int nvoc;

static void feed(llama_context * c, int pos, int n) {
    llama_batch b = llama_batch_init(n, 0, 1);
    for (int i = 0; i < n; i++) {
        b.token[i] = (1000 + pos + i) % nvoc; b.pos[i] = pos + i;
        b.n_seq_id[i] = 1; b.seq_id[i][0] = 0; b.logits[i] = 1;
    }
    b.n_tokens = n;
    llama_decode(c, b);
    llama_batch_free(b);
}

// logits for the token after position WARM, optionally decoding a block there first and
// removing it again -- which must not change anything
static std::vector<float> predict(bool stale, bool fa) {
    llama_context_params cp = llama_context_default_params();
    cp.n_ctx = 256; cp.n_batch = WARM + N + 1;
    cp.flash_attn_type = fa ? LLAMA_FLASH_ATTN_TYPE_ENABLED : LLAMA_FLASH_ATTN_TYPE_DISABLED;
    llama_context * c = llama_init_from_model(model, cp);

    for (int i = 0; i < WARM; i++) feed(c, i, 1);
    if (stale) {
        feed(c, WARM, N);
        llama_memory_seq_rm(llama_get_memory(c), 0, WARM, -1);
    }
    feed(c, WARM, 1);

    const float * lg = llama_get_logits_ith(c, -1);
    std::vector<float> out(lg, lg + nvoc);
    llama_free(c);
    return out;
}

int main(int argc, char ** argv) {
    if (argc != 2) { fprintf(stderr, "usage: %s model.gguf\n", argv[0]); return 2; }
    ggml_backend_load_all();
    llama_backend_init();
    llama_model_params mp = llama_model_default_params();
    mp.n_gpu_layers = 99;
    model = llama_model_load_from_file(argv[1], mp);
    if (!model) return 2;
    if (llama_model_is_recurrent(model) || llama_model_is_hybrid(model)) {
        // rollback on recurrent models is separately broken (#26695) and would confound
        // this test; use a plain transformer
        fprintf(stderr, "needs a non-recurrent model\n"); return 2;
    }
    nvoc = llama_vocab_n_tokens(llama_model_get_vocab(model));

    const bool fa_leaks = predict(false, true)  != predict(true, true);
    const bool no_leaks = predict(false, false) != predict(true, false);

    printf("FA on : %s\nFA off: %s\n%s\n",
           fa_leaks ? "logits CHANGED" : "logits identical",
           no_leaks ? "logits CHANGED" : "logits identical",
           fa_leaks ? "LEAK - freed cells influence the output" : "ok");
    llama_model_free(model);
    llama_backend_free();
    return fa_leaks;
}
```

### First Bad Commit

https://github.com/ggml-org/llama.cpp/commit/aa6f918c1c786668db530c3a1c3ff8a93da928f7

Also see https://github.com/ggml-org/llama.cpp/pull/19625

### Relevant log output

<details>
<summary>Logs</summary>

```console
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = Radeon 8060S Graphics (RADV STRIX_HALO) (radv) | uma: 1 | fp16: 1 | bf16: 0 | fp4: 0 | warp size: 64 | shared memory: 65536 | int dot: 1 | matrix cores: KHR_coopmat
llama_model_loader: loaded meta data with 38 key-value pairs and 290 tensors from /home/nate/models/qwen25-05b/Qwen2.5-0.5B-Instruct-Q8_0.gguf (version GGUF V3 (latest))
print_info: file format = GGUF V3 (latest)
print_info: file type   = Q8_0
print_info: file size   = 500.79 MiB (8.50 BPW) 
print_info: arch                  = qwen2
print_info: n_ctx_train           = 32768
print_info: n_layer               = 24
print_info: n_layer_all           = 24
load_tensors: offloaded 25/25 layers to GPU
llama_context: n_ctx         = 256
llama_context: n_ctx_seq     = 256
llama_context: n_batch       = 15
llama_context: n_ubatch      = 15
llama_context: n_ctx_seq (256) < n_ctx_train (32768) -- the full capacity of the model will not be utilized
llama_kv_cache: size =    3.00 MiB (   256 cells,  24 layers,  1/1 seqs), K (f16):    1.50 MiB, V (f16):    1.50 MiB
FA on : logits CHANGED
FA off: logits identical
LEAK - freed cells influence the output
```
</details>

**Comment by jeffbolznv** [comment]
I haven't been able to reproduce this on either lavapipe or NVIDIA. Maybe a driver bug?

**Comment by natemacfadden** [comment]
I reckon you're right. I tested this on a bunch of different devices/configurations but was unable to recreate it except on my original. See attachments. The "FA pipeline: ..." prints:
```fprintf(stderr, "FA pipeline: %s\n", pipeline->name.c_str());```
in ggml_vk_flash_attn()

[cpu_core_ultra7_270k.log](https://github.com/user-attachments/files/30882129/cpu_core_ultra7_270k.log)
[cpu_i5_10600k.log](https://github.com/user-attachments/files/30882132/cpu_i5_10600k.log)
[cpu_ryzen_ai_max_395.log](https://github.com/user-attachments/files/30882125/cpu_ryzen_ai_max_395.log)
[intel_arl_igpu_scalar.log](https://github.com/user-attachments/files/30882131/intel_arl_igpu_scalar.log)
[radeon8060s_radv_cm1.log](https://github.com/user-attachments/files/30882130/radeon8060s_radv_cm1.log)
[radeon8060s_radv_scalar.log](https://github.com/user-attachments/files/30882133/radeon8060s_radv_scalar.log)
[rtx5060ti_nvidia_cm1_forced.log](https://github.com/user-attachments/files/30882126/rtx5060ti_nvidia_cm1_forced.log)
[rtx5060ti_nvidia_cm2.log](https://github.com/user-attachments/files/30882128/rtx5060ti_nvidia_cm2.log)
[rx6700xt_radv_scalar.log](https://github.com/user-attachments/files/30882127/rx6700xt_radv_scalar.log)

**Comment by natemacfadden** [comment]
OK really not a llama bug - I can close this issue if desired. I recreated this outside llama.cpp with a single 16x16 coopMatMulAdd on RADV/RDNA3.



I'll detail what I did below:
-----------------------------
Let A be a 16x16 matrix that is +0 except one column of -1.0. Let B be a 16x16 matrix of all +0 except one row of 0.0625, 0.125, ..., 1.0. We compute C = A * B.

One way, we do it for A,B listed above. The other way, flip one all-(+0) row of B to all-(-0) (i.e., change the sign). These two prescriptions give slightly different outputs. The discrepancy is small though (~1 ulp) so IDK how much people care about this.

I only saw this on the f32-accumulator wmma. I also didn't see it on my 5060Ti


(the above test is somewhat analogous to i) a normal decode and ii) a decode after undoing a draft and masked-out KV cache messes up some sums. that's how I encountered it)
